cmd/soroban-cli/src/events.rs

```diff
@@ -0,0 +1,192 @@
+    start_ledger: u32,
+    end_ledger: u32,
+
+    /// RPC server endpoint
+    #[clap(long,
+        env = "SOROBAN_RPC_URL",
+        help_heading = HEADING_RPC,
+        conflicts_with = "sandbox",
+    )]
+    rpc_url: Option<String>,
+
+    /// Sandbox file from which to pull events from
+    #[clap(
+        long,
+        parse(from_os_str),
+        value_name = "PATH",
+        env = "SOROBAN_LEDGER_FILE",
+        help_heading = HEADING_SANDBOX,
+        conflicts_with = "rpc-url",
+    )]
+    sandbox: Option<std::path::PathBuf>,
+
+    /// A set of (up to 5) contract IDs to filter events on
+    #[clap(long = "ids", multiple = true)]
+    contract_ids: Vec<String>,
+
+    /// A set of (up to 5) topic filters to filter events on
+    #[clap(long, multiple = true)]
+    topics: Vec<String>,
+
+    /// Formatting options for outputted events
+    #[clap(
+        long,
+        possible_values = ["console", "json"],
+        default_value = "console"
+    )]
+    format: String,
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum Error {
+    #[error("invalid ledger range: low bigger than high ({low} > {high})")]
+    InvalidLedgerRange { low: u32, high: u32 },
+
+    #[error("cannot parse contract ID {contract_id}: {error}")]
+    CannotParseContractId {
+        contract_id: String,
+        error: FromHexError,
+    },
+
+    #[error("too many contracts IDs (max 5)")]
+    TooManyContractIds {},
+
+    #[error("too many topic filters (max 5)")]
+    TooManyTopicFilters {},
+
+    #[error("you must specify either an RPC server or sandbox filepath")]
+    TargetRequired {},
+
+    #[error("sandbox filepath does not exist: {path}")]
+    InvalidSandboxFile { path: String },
+
+    #[error(transparent)]
+    Rpc(#[from] rpc::Error),
+
+    #[error(transparent)]
+    Generic(#[from] Box<dyn std::error::Error>),
+}
+
+impl Cmd {
+    pub async fn run(&self, _matches: &clap::ArgMatches) -> Result<(), Error> {
+        if self.contract_ids.len() > 5 {
+            return Err(Error::TooManyContractIds {});
+        }
+
+        if self.topics.len() > 5 {
+            return Err(Error::TooManyTopicFilters {});
+        }
+
+        if self.start_ledger > self.end_ledger {
+            return Err(Error::InvalidLedgerRange {
+                low: self.start_ledger,
+                high: self.end_ledger,
+            });
+        }
+
+        for raw_contract_id in &self.contract_ids {
+            // We parse the contract IDs to ensure they're the correct format,
+            // but since we'll be passing them as-is to the RPC server anyway,
+            // we disregard the return value.
+            utils::contract_id_from_str(raw_contract_id).map_err(|e| {
+                Error::CannotParseContractId {
+                    contract_id: raw_contract_id.clone(),
+                    error: e,
+                }
+            })?;
+        }
+
+        let mut events: Vec<Event> = Vec::new();
+        if self.rpc_url.is_some() {
```

Comment:
if let is not stable so this could be:

```rust
if let Some(rpc_url) = self.rpc_url.as_ref() {
```

Which removes the need to unwrap in the branch. Same goes for the next case.

Reply:
Nice! Done :+1: